File: recruiter-outreach-ui/src/services/emailSenderApi.ts
import { ApiEndpoints, Messages, Values } from '../constants';

export type TemplateVariant = {
  subjectTemplate: string;
  bodyTemplate: string;
};

export type TemplateDto = {
  key: string;
  displayName: string;
  availableTemplates: string[];
  templates?: {
    hr?: TemplateVariant;
    referral?: TemplateVariant;
  };
};

export async function fetchTemplates(): Promise<TemplateDto[]> {
  const response = await fetch(ApiEndpoints.Templates);
  if (!response.ok) {
    throw new Error(Messages.LoadTemplatesStatus(response.status));
  }

  const raw = (await response.json()) as any[];

  const data: TemplateDto[] = raw.map((r) => ({
    key: r.key,
    displayName: r.displayName,
    availableTemplates: (r.availableTemplates ?? []) as string[],
    templates: r.templates
      ? {
          hr: r.templates.hr
            ? {
                subjectTemplate:
                  r.templates.hr.subjectTemplate ?? r.templates.hr.SubjectTemplate ?? '',
                bodyTemplate:
                  r.templates.hr.bodyTemplate ?? r.templates.hr.BodyTemplate ?? '',
              }
            : undefined,
          referral: r.templates.referral
            ? {
                subjectTemplate:
                  r.templates.referral.subjectTemplate ?? r.templates.referral.SubjectTemplate ?? '',
                bodyTemplate:
                  r.templates.referral.bodyTemplate ?? r.templates.referral.BodyTemplate ?? '',
              }
            : undefined,
        }
      : undefined,
  }));

  return data;
}

export async function sendEmails(input: {
  recruiters: string;
  roleKey: string;
  templateKind: string;
  subject: string;
  emailBody: string;
  resumeFile?: File | null;
  persistResume: boolean;
}): Promise<void> {
  const formData = new FormData();
  formData.append('recruiters', input.recruiters);
  formData.append('roleKey', input.roleKey);
  formData.append('templateKind', input.templateKind);
  formData.append('company', Values.EmptyCompany);
  formData.append('subject', input.subject);
  formData.append('emailBody', input.emailBody);
  formData.append('persistResume', String(input.persistResume));

  if (input.resumeFile) {
    formData.append('resume', input.resumeFile);
  }

  const response = await fetch(ApiEndpoints.SendEmails, {
    method: 'POST',
    body: formData,
  });

  if (!response.ok) {
    if (response.status === 401) {
      const err: any = new Error(Messages.SessionExpired);
      err.code = Values.SessionExpiredCode;
      err.status = 401;
      throw err;
    }
    const text = await response.text().catch(() => '');
    throw new Error(text || Messages.SendStatus(response.status));
  }
}

File: RecruiterOutreach.Api/Endpoints/EmailSenderApi.cs
using System.Text.RegularExpressions;
using Microsoft.AspNetCore.Mvc;
using RecruiterOutreach.Core;
using RecruiterOutreach.Core.Outreach;

namespace RecruiterOutreach.Api.Endpoints;

/// <summary>
/// Maps email-sending and template-related endpoints used by the email sender tab.
/// </summary>
public static class EmailSenderApi
{
    public static IEndpointRouteBuilder MapEmailSenderApi(this IEndpointRouteBuilder app)
    {
        // /api/templates
        app.MapGet(Constants.ApiEndpoints.GetTemplatesApi, (OutreachSettings settings) =>
        {
            var roleTemplates = settings.RoleEmailTemplates ?? new List<RoleEmailTemplateSettings>();

            var roles = roleTemplates
                .Select(r => new
                {
                    r.Key,
                    r.DisplayName,
                    AvailableTemplates = (r.Templates is null
                        ? Array.Empty<string>()
                        : new[]
                        {
                            r.Templates.Hr is not null ? Constants.Defaults.Hr : null,
                            r.Templates.Referral is not null ? Constants.Defaults.Referral : null
                        }.Where(x => x is not null)),
                    Templates = r.Templates is null
                        ? null
                        : new
                        {
                            hr = r.Templates.Hr is not null
                                ? new { r.Templates.Hr.SubjectTemplate, r.Templates.Hr.BodyTemplate }
                                : null,
                            referral = r.Templates.Referral is not null
                                ? new { r.Templates.Referral.SubjectTemplate, r.Templates.Referral.BodyTemplate }
                                : null
                        }
                });

            return Results.Ok(roles);
        })
        .WithName(Constants.Defaults.Templates);

        // /api/templates/preview
        app.MapGet(Constants.ApiEndpoints.GetTemplatePreviewApi, ([FromQuery] string roleKey, [FromQuery] string templateKind, OutreachSettings settings) =>
        {
            var role = settings.RoleEmailTemplates
                .FirstOrDefault(r => string.Equals(r.Key, roleKey, StringComparison.OrdinalIgnoreCase));

            if (role is null)
            {
                return Results.NotFound();
            }

            var variant = templateKind.Equals(Constants.Defaults.Referral, StringComparison.OrdinalIgnoreCase)
                ? role.Templates.Referral
                : role.Templates.Hr;

            if (variant is null)
            {
                return Results.NotFound();
            }

            return Results.Ok(new
            {
                variant.SubjectTemplate,
                variant.BodyTemplate
            });
        })
        .WithName(Constants.Defaults.GetTemplatePreview);

        // /api/outreach/send
        app.MapPost(Constants.ApiEndpoints.SendEmailApi, async (
            [FromForm] string recruiters,
            [FromForm] string? roleKey,
            [FromForm] string? templateKind,
            [FromForm] string? company,
            [FromForm] string? subject,
            [FromForm] string? emailBody,
            [FromForm] bool persistResume,
            IFormFile? resume,
            OutreachService outreach,
            OutreachSettings settings) =>
        {
            // Allow recruiter emails to be separated by commas, semicolons, spaces, or newlines
            var recruiterEmails = Regex.Split(recruiters ?? string.Empty, Constants.Email.EmailSplitPattern)
                .Where(x => !string.IsNullOrWhiteSpace(x))
                .ToArray();

            company ??= string.Empty;

            // If a resume is uploaded, either persist it as the new default (when requested) or
            // write it to a temporary file that will be used only for this send.
            string? resumeAttachmentPath = null;
            if (resume is not null && resume.Length > 0)
            {
                if (persistResume && settings.Resume is not null && !string.IsNullOrWhiteSpace(settings.Resume.DefaultAttachmentPath))
                {
                    var targetPath = settings.Resume.DefaultAttachmentPath;
                    Directory.CreateDirectory(Path.GetDirectoryName(targetPath)!);
                    using (var fileStream = File.Create(targetPath))
                    {
                        await resume.CopyToAsync(fileStream);
                    }

                    resumeAttachmentPath = targetPath;
                }
                else
                {
                    var baseFolder = !string.IsNullOrWhiteSpace(settings.Resume?.DefaultAttachmentPath)
                        ? Path.GetDirectoryName(settings.Resume.DefaultAttachmentPath) ?? Directory.GetCurrentDirectory()
                        : Path.GetDirectoryName(Environment.ProcessPath!) ?? Directory.GetCurrentDirectory();

                    Directory.CreateDirectory(baseFolder);

                    var tempFileName = Constants.Defaults.GenerateResumeFileName(resume.FileName);
                    var tempPath = Path.Combine(baseFolder, tempFileName);

                    using (var fileStream = File.Create(tempPath))
                    {
                        await resume.CopyToAsync(fileStream);
                    }

                    resumeAttachmentPath = tempPath;
                }
            }

            // Resolve a human-friendly role name for token replacement (e.g., "Data Analyst")
            var roleDisplayName = string.Empty;
            if (!string.IsNullOrWhiteSpace(roleKey))
            {
                var role = settings.RoleEmailTemplates
                    .FirstOrDefault(r => string.Equals(r.Key, roleKey, StringComparison.OrdinalIgnoreCase));

                // Prefer DisplayName, but fall back to the key itself so {Role} is never empty
                roleDisplayName = !string.IsNullOrWhiteSpace(role?.DisplayName)
                    ? role!.DisplayName
                    : roleKey;
            }

            await outreach.RunAsync(
                company,
                roleDisplayName,
                jobDescription: null,
                roleKey: roleKey,
                templateKind: templateKind,
                recruiterEmails: recruiterEmails,
                subjectOverride: subject,
                emailBodyOverride: emailBody,
                resumeAttachmentPath: resumeAttachmentPath);

            return Results.Ok();
        })
        .DisableAntiforgery()
        .WithName(Constants.Defaults.SendEmails);

        return app;
    }
}
